verilog/tools/kythe/scope_resolver.h

```diff
@@ -0,0 +1,199 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_TOOLS_KYTHE_SCOPE_RESOLVER_H_
+#define VERIBLE_VERILOG_TOOLS_KYTHE_SCOPE_RESOLVER_H_
+
+#include <map>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+#include "common/util/auto_pop_stack.h"
+#include "common/util/iterator_range.h"
+#include "verilog/tools/kythe/kythe_facts.h"
+
+namespace verilog {
+namespace kythe {
+
+// Used to wrap whatever needs to be recorded in a scope item.
+struct ScopeMemberItem {
+  ScopeMemberItem(const VName& vname) : vname(vname) {}
+
+  // VName of this member.
+  VName vname;
+};
+
+// Represents some scope in SystemVerilog code like class or module.
+class Scope {
+ public:
+  Scope() {}
+  Scope(const Signature& signature) : signature_(signature) {}
+
+  // Appends the given VName to the members of this scope.
+  void AddMemberItem(const ScopeMemberItem& vname);
+
+  // Appends the member of the given scope to the current scope.
+  void AppendScope(const Scope& scope);
+
+  const std::vector<ScopeMemberItem>& Members() const { return members_; }
+  const Signature& GetSignature() const { return signature_; }
+
+  // Searches for the given reference_name in the current scope and returns its
+  // VName or nullptr if not found.
+  const VName* SearchForDefinition(absl::string_view name) const;
+
+ private:
+  // Signature of the owner of this scope.
+  Signature signature_;
+
+  // list of the members inside this scope.
+  std::vector<ScopeMemberItem> members_;
+};
+
+// Container with a stack of Scopes to hold the accessible scopes during
+// traversing an Indexing Facts Tree.
+// This is used to get the definitions of some reference.
+//
+// This is modified during tree traversal because in case of entering new
+// scope the new scope is resolved first and after that it's added to the
+// containing scope and the next scope is being analyzed.
+// e.g
+// package pkg;
+//    int x;
+//    function int my_fun();
+//      return x;
+//    endfunction
+// endpackage
+//
+// Vertical scope when extracting my_fun:
+// The Global Scope
+// {
+//    "pkg"
+// }
+// Pkg Scope
+// {
+//   pkg#x
+// }
+// when trying to find a definition for "x" in "return x" VerticalScopeResolver
+// searches the above scope in reverse order till if finds a definition or
+// returns a nullptr if not found.
+class VerticalScopeResolver : public verible::AutoPopStack<Scope*> {
+ public:
+  typedef verible::AutoPopStack<Scope*> base_type;
+
+  // member class to handle push and pop of stack safely
+  using AutoPop = base_type::AutoPop;
+
+  // returns the top VName of the stack
+  Scope& top() { return *ABSL_DIE_IF_NULL(base_type::top()); }
```

Comment:
Is there a good use for a `const` variant of this to allow non-modifying access?

```c++
  const Scope& top() const { return *ABSL_DIE_IF_NULL(base_type::top()); }
```

Reply:
For now this scope is modified a lot while traversing.